### ISSUE TITLE
Major Performance Issue in GlyphBrush::draw_queued

### ISSUE THREAD

**Issue by jojolepro** [alexheretic/glyph-brush#52]
Using thread_profiler, I tracked down a performance issue text redraws of changing text to take over 30ms (and in some cases 250!).

The line in question is this one: https://github.com/alexheretic/glyph-brush/blob/0a72b6e071463c5b4dbbff05fd74693e29d0fbb9/gfx-glyph/src/lib.rs#L384

I can provide the version with the profiler code if you need it.

Thanks!

**Comment by alexheretic** [comment]
Interesting, though this may be hard to deal with as it could be hardware/driver specific. This line is uploading the vertices to the gpu.

I haven't noticed this being a bottleneck myself though. Can you test on your hardware with the performance example (that has ~30000 unicode glyphs to render)?

### Apply patch
```
diff --git a/gfx-glyph/src/lib.rs b/gfx-glyph/src/lib.rs
index 68d0d8b..d6ba9e4 100644
--- a/gfx-glyph/src/lib.rs
+++ b/gfx-glyph/src/lib.rs
@@ -381,7 +381,9 @@ impl<'font, R: gfx::Resources, F: gfx::Factory<R>, H: BuildHasher> GlyphBrush<'f
 
         match brush_action.unwrap() {
             BrushAction::Draw(verts) => {
+                let before_vbuf = std::time::Instant::now();
                 let vbuf = self.factory.create_vertex_buffer(&verts);
+                warn!("vbuf creation took {:.1?}", before_vbuf.elapsed());
 
                 let draw_cache = if let Some(mut cache) = self.draw_cache.take() {
                     cache.pipe_data.vbuf = vbuf;
```
### Run example
`RUST_LOG=warn cargo run --example performance --release`

It'll log each time it regenerates vertices. Scrolling out forces it to generate all vertices, on my setup this line logs `[WARN  gfx_glyph] vbuf creation took 678.9µs` as a worst case.

### Caching
As a more general note on the caching behaviour of glyph-brush, best performance is achieved by minimising text changes. So, for example, you may get better results by moving some distinct text to a separate glyph brush to isolate the caching and not force full vertex generation for any text change.

It would be interesting to see if it's possible to have a vertex buffer per section so each section had an isolated cache of sorts. This would need investigating with gfx pre-ll.

**Comment by jojolepro** [comment]
I forgot to specify this happens only windows, sorry ;)

**Comment by jojolepro** [comment]
I'm doing more testing, the profiling results are inconsistent now...

**Comment by alexheretic** [comment]
Testing the performance example in windows it actually seems to upload vertices a little faster than linux...

```
[WARN  gfx_glyph] create_vertex_buffer(v x 19549) in 96.1µs
[WARN  gfx_glyph] create_vertex_buffer(v x 30001) in 152.4µs
[WARN  gfx_glyph] create_vertex_buffer(v x 30001) in 152.1µs
[WARN  gfx_glyph] create_vertex_buffer(v x 21999) in 119.6µs
[WARN  gfx_glyph] create_vertex_buffer(v x 30001) in 151.8µs
```

**Comment by alexheretic** [comment]
I'm using a rx480/amd drivers, and this is probably driver specific. Are there better opengl calls to make here? It's difficult to see how to improve this easily, particularly as I'm having trouble reproducing it on my setup.

**Comment by jojolepro** [comment]
Here's the profiling of the performance example.
https://www.dropbox.com/s/fknqnadcppujfk3/thread_profiler.json?dl=1
You can open it in chrome by going to this uri: chrome://tracing

The first part is with my laptop unplugged, and the second one is after plugging the power cord.
When I was typing, you can clearly see big spikes in the execution time whenever I would type new text.

I also tested the "varied" example. When I was resizing, the framerate would drop to 30.

All of this is pretty much expected however, so nothing to be fixed here.

However, when I test amethyst, with the laptop plugged, it actually does have an issue with uploading the vertex buffer. I think this might be a problem on our side, but I can't quite figure out why.

This is the profiling when running the amethyst "ui" example.
https://www.dropbox.com/s/6gl1cssioyr02dg/thread_profile_amethyst.json?dl=1

The code in question is here: https://github.com/amethyst/amethyst/blob/master/amethyst_ui/src/pass.rs

Maybe we improperly setup caching, but I'm not sure how that would make the vertex buffer upload time so high. I'm on nvidia, so maybe it is a vendor-specific issue?

**Comment by alexheretic** [comment]
Yes I haven't seen vertex upload being an issue. Changing text causing performance spikes is expected, but it's normally CPU side. ie the cost of redoing layout, texture cache calc & vertex generation. I'd expect something like this running performance example:
```
[WARN  glyph_brush::glyph_brush] section layout(t.len()=70078) in 12.0ms
[WARN  glyph_brush::glyph_brush] texture cache processing in 1.6ms
[WARN  glyph_brush::glyph_brush] vertex generation 5.3ms
[WARN  gfx_glyph] create_vertex_buffer(v x 30001) in 216.5µs
```
In this case the layout work & vertex generation are much slower than the actual upload call. And this is generally what I've been trying to improve, the CPU side of it. And I believe there **is** more I can do here to improve glyph-brush performance.

But I'm less sure about the vertex upload in gfx-glyph. It would be nice to be able to do partial uploads when not every vertex has changed, but this seems kind of difficult. It would definitely be nice to improve the upload as it's fairly brutish right now.

**Comment by jojolepro** [comment]
Following tests made by the community here https://github.com/amethyst/amethyst/issues/1097
it seems that this only affects nvidia gpu on a consistent basis.

We are not sure if that would be a driver bug on them, or if the usage this is a bug in gfx_window_glutin, or if it is cause by something in glyph-brush.

Any idea? 
Thanks!

**Comment by alexheretic** [comment]
Maybe there's an optimisation for nvidia to be made in gfx_device_gl: https://github.com/gfx-rs/gfx/blob/de15943595d3a3d95b8b00595888e3483d01e924/src/backend/gl/src/factory.rs#L329 

Can you trace this down to slow GL call(s), that's where I'd guess the issue is.

**Comment by alexheretic** [comment]
We can also play around with the opengl buffer data methods, though let me say now I'm not really a opengl expert.

Can you reproduce the slow upload with the glyph-brush opengl example? It's much closer to the opengl API, so it's easier to play with.

### Use partial updates & stream_draw
* The example uses `BufferData` to upload the vertices. It _could_ use that only when the buffer needs allocation, instead using `BufferSubData` to upload the vertices. This should be faster.
* The opengl example also reflect the gfx-glyph gfx calls in that it will use `STATIC_DRAW`. This probably isn't the best fit, it makes sense (particularly when doing partial updates often) to use `STREAM_DRAW`.

```patch
diff --git a/glyph-brush/examples/opengl.rs b/glyph-brush/examples/opengl.rs
index 2222ab8..ae4e527 100644
--- a/glyph-brush/examples/opengl.rs
+++ b/glyph-brush/examples/opengl.rs
@@ -151,6 +151,7 @@ fn main() -> Res<()> {
     let mut loop_helper = spin_sleep::LoopHelper::builder().build_with_target_rate(250.0);
     let mut running = true;
     let mut vertex_count = 0;
+    let mut vertex_max = vertex_count;
     let mut dimensions = window
         .get_inner_size()
         .ok_or("get_inner_size = None")?
@@ -304,14 +305,27 @@ fn main() -> Res<()> {
             BrushAction::Draw(vertices) => {
                 // Draw new vertices
                 vertex_count = vertices.len();
+                let before = std::time::Instant::now();
                 unsafe {
-                    gl::BufferData(
-                        gl::ARRAY_BUFFER,
-                        (vertex_count * mem::size_of::<Vertex>()) as GLsizeiptr,
-                        vertices.as_ptr() as _,
-                        gl::STATIC_DRAW,
-                    );
+                    if vertex_max < vertex_count {
+                        gl::BufferData(
+                            gl::ARRAY_BUFFER,
+                            (vertex_count * mem::size_of::<Vertex>()) as GLsizeiptr,
+                            vertices.as_ptr() as _,
+                            gl::STREAM_DRAW,
+                        );
+                        dbg!(("BufferData", before.elapsed()));
+                    } else {
+                        gl::BufferSubData(
+                            gl::ARRAY_BUFFER,
+                            0,
+                            (vertex_count * mem::size_of::<Vertex>()) as GLsizeiptr,
+                            vertices.as_ptr() as _,
+                        );
+                        dbg!(("BufferSubData", before.elapsed()));
+                    }
                 }
+                vertex_max = vertex_max.max(vertex_count);
             }
             BrushAction::ReDraw => {}
         }
```

I'll try measuring myself if these could improve the upload time. Because if I can measure it, even if it's a small improvement, it might help more for the suffering nvidia case.

**Comment by alexheretic** [comment]
Ok yes, I think this is an optimisation for everyone. I've switched to `DYNAMIC_DRAW` as gfx didn't like stream/upload. Though my worst case upload speed is quite fast, this does seem to improve it, can you retest with commit 9029e26175efed94992c8867977b52db985fc53d?

I don't have any formal benchmarks for the opengl calls in gfx-glyph, as I'm generally looking at glyph-brush cpu performance.

**Comment by jojolepro** [comment]
Sorry for the delay, I had a busy week ^^

Good news though, your patch worked!
As it was said today on the amethyst discord, the best guess is that creating a vertex buffer this often over saturates the nvidia driver *for some reason*.

Would you mind releasing this as a patch or minor version change to crates.io?

Thanks a lot! <3 <3 <3 <3

**Comment by alexheretic** [comment]
That's great news, I'm glad this optimisation is enough to cure your issue! 

Thanks for raising this, as I'll be able to apply this knowhow to quad rendering in Robo Instructus.

I'm planning to release the queued changes this Friday.

**Comment by alexheretic** [comment]
This optimisation is available in gfx-glyph `0.13.2`.
